### PR TITLE
Port multilateration

### PR DESCRIPTION
**Pull request by hannahcl**
Port multilateration code from python to cpp. Test that it gives same result. 
[Issue](https://github.com/vortexntnu/acoustics/issues/72)

### CODE REVIEW

File: cpp/teensy/lib/multilateration/multilateration.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "arm_math.h"
+#include <Arduino.h>
+#include <cassert>
+
+#ifndef HYDROPHONE_DETAILES
+#define HYDROPHONE_DETAILES
```
Comment (nicbres) [inline cpp/teensy/lib/multilateration/multilateration.h:8]:
```suggestion
#ifndef HYDROPHONE_DETAILS
#define HYDROPHONE_DETAILS
```

File: cpp/teensy/lib/multilateration/multilateration.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+#endif
+
+#ifndef DSP_CONSTANTS
+#define DSP_CONSTANTS
+
+#define SAMPLING_FREQ 300000             //[Hz]
+#define SIGNAL_LENGTH 40 * SAMPLING_FREQ //[sampels]
+
+#endif
```
Comment (nicbres) [inline cpp/teensy/lib/multilateration/multilateration.h:48]: I think we should replace all of these with `const` variables instead of using defines.

File: cpp/teensy/lib/multilateration/multilateration.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
+#include "multilateration.h"
+
+/*
+TODO:
+- clean up main (delete all)
+
+*/
```
Comment (nicbres) [inline cpp/teensy/lib/multilateration/multilateration.cpp:7]:
```suggestion
```
Cleanup the cleanup comment (:

File: cpp/teensy/lib/multilateration/multilateration.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+struct HydrophonePositions {
+    float32_t PosX;
+    float32_t PosY;
+    float32_t PosZ;
+};
```
Comment (nicbres) [inline cpp/teensy/lib/multilateration/multilateration.h:54]:
```suggestion
struct Position {
    float32_t X;
    float32_t Y;
    float32_t Z;
};
```
Redefining this struct we can also use it as the result in the `calculatePingerPosition`.

File: cpp/teensy/lib/multilateration/multilateration.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    HydrophonePositions Hyd2 = {HYD_2_X_POS, HYD_2_Y_POS, HYD_2_Z_POS};
+    HydrophonePositions Hyd3 = {HYD_3_X_POS, HYD_3_Y_POS, HYD_3_Z_POS};
+    HydrophonePositions Hyd4 = {HYD_4_X_POS, HYD_4_Y_POS, HYD_4_Z_POS};
+
+    hydrophonePositions[0] = Hyd0;
+    hydrophonePositions[1] = Hyd1;
+    hydrophonePositions[2] = Hyd2;
+    hydrophonePositions[3] = Hyd3;
+    hydrophonePositions[4] = Hyd4;
+}
```
Comment (nicbres) [inline cpp/teensy/lib/multilateration/multilateration.cpp:125]: Should probably be a constant array of `const Position HydrophonePositions[5]`.

File: cpp/teensy/lib/multilateration/multilateration.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+TODO:
+- clean up main (delete all)
+
+*/
+
+arm_status calculatePingerPosition(int32_t TdoaArray[],
+                                   HydrophonePositions HydrophonePositions[],
+                                   const arm_matrix_instance_f32* pA,
+                                   const arm_matrix_instance_f32* pB,
+                                   arm_matrix_instance_f32* pResult) {
```
Comment (nicbres) [inline cpp/teensy/lib/multilateration/multilateration.cpp:13]: IMO the result of this function should be a position. 

```suggestion
arm_status calculatePingerPosition(int32_t TdoaArray[],
                                   HydrophonePositions HydrophonePositions[],
                                   const arm_matrix_instance_f32* pA,
                                   const arm_matrix_instance_f32* pB,
                                   Position* pResult) {
```

File: cpp/teensy/test/test_multilateration/test_multilateration.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
+#include "Arduino.h"
+#include "multilateration.h"
+#include "stdio.h"
```
Comment (nicbres) [inline cpp/teensy/test/test_multilateration/test_multilateration.cpp:3]:
```suggestion
```
Why is this here?

File: Python/multilateration/test/test_multilateration.py [hunk 8]
```diff
@@ -131,3 +132,52 @@ def test_trilateration_algorithm():
...
+        ]
+    )
+
+    x_estimate, y_estimate, z_estimate = mult.calculate_pinger_position(
+        tdoa_lag_array=tdoa_sample_array,
+        hydrophone_positions=hydrophone_positions,
+        sample_frequency=sample_frequency,
+    )
+
+    print(f"x: {x_estimate}, y: {y_estimate}, z: {z_estimate} ")
```
[the remark this answers is not in the document]
Comment (hannahcl) [reply Python/multilateration/test/test_multilateration.py:187]: I agree, but actually in this test it is need in order to compare the results to what you get on the teensy.

File: cpp/teensy/lib/multilateration/multilateration.h (hunk 4 above)
Comment (hannahcl) [reply cpp/teensy/lib/multilateration/multilateration.h:54]: What do you think now?

File: cpp/teensy/lib/multilateration/multilateration.cpp (hunk 5 above)
Comment (hannahcl) [reply cpp/teensy/lib/multilateration/multilateration.cpp:125]: What would be a good way to do this? I could allocate an array inside the init function, and use a return value, but this is not very pretty.

Comment (nicbres) [reply cpp/teensy/lib/multilateration/multilateration.cpp]: This is a constant at this point.
```C++
const Position hydrophonePositions[5] = {
    {..., ..., ...},
    .    ...
    {..., ..., ...},
};
```
Is there a reason this doesn't work?

File: cpp/teensy/lib/multilateration/multilateration.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+#endif
+
+#ifndef DSP_CONSTANTS
+#define DSP_CONSTANTS
+
+const int SAMPLING_FREQ = 300000; //[Hz]
+
+#endif
+
+struct Positions {
```
Comment (nicbres) [inline cpp/teensy/lib/multilateration/multilateration.h:48]:
```suggestion
struct Position {
```
It only becomes plural in an array ;)

File: cpp/teensy/lib/multilateration/multilateration.h [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+const float32_t HYD_2_Y_POS = -0.24;
+const float32_t HYD_2_Z_POS = 0.0;
+
+const float32_t HYD_3_X_POS = 0.5;
+const float32_t HYD_3_Y_POS = -0.1;
+const float32_t HYD_3_Z_POS = 0.4;
+
+const float32_t HYD_4_X_POS = 0.4;
+const float32_t HYD_4_Y_POS = 0.0;
+const float32_t HYD_4_Z_POS = -0.4;
```
Comment (nicbres) [inline cpp/teensy/lib/multilateration/multilateration.h:30]: If you use the `const Position hydrophonePositions[5]` thing, I'd say all of these are mostly obsolete.

File: cpp/teensy/lib/multilateration/multilateration.cpp [hunk 11]
```diff
@@ -0,0 +1,128 @@
+#include "multilateration.h"
+
+arm_status calculatePingerPosition(int32_t TdoaArray[],
+                                   Positions hydrophonePositions[],
+                                   const arm_matrix_instance_f32* pA,
+                                   const arm_matrix_instance_f32* pB,
+                                   arm_matrix_instance_f32* pResult,
```
[the remark this answers is not in the document]
Comment (hannahcl) [reply cpp/teensy/lib/multilateration/multilateration.cpp:7]: Do you think this is better?

**Review by nicbres — APPROVED** [review]
LGTM
